app/src/main/java/com/duckduckgo/app/browser/downloader/DataUriParser.kt

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2018 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.browser.downloader
+
+import com.duckduckgo.app.browser.downloader.DataUriParser.ParseResult.ParsedDataUri
+import timber.log.Timber
+import java.util.*
+import javax.inject.Inject
+
+class DataUriParser @Inject constructor() {
+
+    fun generate(url: String): ParseResult {
+        val result = MIME_TYPE_REGEX.find(url)
+        if (result == null || result.groupValues.size < 5) {
+            return ParseResult.Invalid
+        }
+
+        val mimeType = result.groupValues[REGEX_GROUP_MIMETYPE]
+        val fileTypeGeneral = result.groupValues[REGEX_GROUP_FILE_TYPE_GENERAL]
+        val fileTypeSpecific = result.groupValues[REGEX_GROUP_FILE_TYPE_SPECIFIC]
+        val data = result.groupValues[REGEX_GROUP_DATA]
+
+        val suffix = determineSuffix(fileTypeGeneral, fileTypeSpecific)
+        val filename = UUID.randomUUID().toString()
+        val generatedFilename = GeneratedFilename(name = filename, fileType = suffix)
+
+        return ParsedDataUri(fileTypeGeneral, fileTypeSpecific, data, mimeType, generatedFilename)
+    }
+
+    private fun determineSuffix(fileTypeGeneral: String, fileTypeSpecific: String): String {
+        return when (fileTypeGeneral) {
```

Comment:
Could use `MimeTypeMap` here? 

See https://developer.android.com/reference/android/webkit/MimeTypeMap.html#getExtensionFromMimeType(java.lang.String)

Reply:
Inspired! I looked for this and failed to find it. And it fixes every other piece of feedback you've raised too! 🙏